fdbserver/workloads/FastTriggeredWatches.actor.cpp

```diff
@@ -145,7 +145,7 @@ struct FastTriggeredWatchesWorkload : TestWorkload {
 				//TraceEvent("FTWWatchDone").detail("Key", printable(setKey));
 				// Assert that the time from setting the key to triggering the watch is no greater than 25s
 				// TODO: This assertion can cause flaky behaviour since sometimes a watch can take longer to fire
-				ASSERT(lastReadVersion - ver >= SERVER_KNOBS->MAX_VERSIONS_IN_FLIGHT ||
+				ASSERT(first || lastReadVersion - ver >= SERVER_KNOBS->MAX_VERSIONS_IN_FLIGHT ||
```

Comment:
nice catch!

Reply:
In a test run with additional printing, I got
```
first = true lastReadVersion=441719868 ver=409158709 max_in_flight=100000000 versions_per_s=1000000 duration=0.00
```

Reply:
It's possible for the lastReadVersion to be greater if the `setter commits its transaction before we start the transaction for the watch

Reply:
I'm pretty sure this fix would not fix all occurrences of this test being flaky if for example we hit this case where first is false but due to attrition the lastReadVersion >> ver

Reply:
I'm glad we are on this topic :)
I noticed that in the same file, but earlier, we have
```
void disableFailureInjectionWorkloads(std::set<std::string>& out) const override {
		// This test asserts that watches fire within a certain version range. Attrition will make this assertion fail
		// since it can cause recoveries which will bump the cluster version significantly
		out.emplace("Attrition");
	}
```
I wonder if the attrition case has already been accounted for.

Reply:
Hmm interesting im not really sure then what could be causing the versions to change dramatically. Maybe there's some BUGGIFY logic somewhere?

Reply:
Yeah, the version can be greater (as it is normally) or lesser (should be unique to this case), but if greater I don't really understand why it would be any worse than the normal case. It's still not clear to me what this actually fixes related to the assert (though I do agree with the idea to avoid generating a value that bypasses the test).

Reply:
Related to the recoveries bumping versions, it seems as if the test is at least trying to account for that by filtering out version differences larger than `MAX_VERSIONS_IN_FLIGHT`

Reply:
Ah, I guess maybe what's different is we are susceptible to a long delay in getting a read version and reading the key on the first pass? This is I assume the problem you were alluding to when you mentioned that `getDuration` being 0 was a problem.

Interestingly, I think this may be a problem even if we guarantee use of different keys. If the sequence looks like this:

1. Set key commits at version X
2. Transaction takes a long time to get a read version and read, say 30s
3. We read the value. Either this happens at a version Y>=X, in which case we see the same value and break the loop (triggering the existing assert, fixed by the new assert), or it somehow happens at Y < X (i.e. we get a read version and then have a very long delay for reading the key, which I think is possible but should fail to commit).

Avoiding checking the elapsed versions if we didn't actually set the watch seems like a reasonable choice, then.

Reply:
> Avoiding checking the elapsed versions if we didn't actually set the watch seems like a reasonable choice, then.

iiuc, the current fix uses the `first` variable to indicate if we actually set the watch. Is this right?

Reply:
Yeah, exactly

Reply:
> happens at Y < X (i.e. we get a read version and then have a very long delay for reading the key, which I think is possible but should fail to commit

Oh, the transaction trying to set a watch will fail to commit because it's not read-only? If so, it will actually retry with `first` remaining true which falls back to the case of `Y >= X`. Correct?

Reply:
Actually, maybe it won't fail on commit because it is read-only. And if that happens, then we are susceptible to this even on the second pass (where `first == false`) because our timing of the watch triggering includes all of the time that we spent getting the read version initially.

Reply:
That sequence would look like:

1. Get a read version X
2. Set the key at version Y = X+n
3. Long delay (say 30s)
4. Read the key from a SS that is lagging and can still serve old reads
5. Set the watch and commit it
6. Watch fires
7. Re-read the key at version Z>Y+30s
8. Fail our assert because `first == false`

Reply:
I think a simple fix for this would be to force the transaction to commit (e.g. by setting a write conflict range). We could even then incorporate the commit version into our timing as a way to eliminate the time spent in the first transaction from our timing:

int64_t versionDelta = lastReadVersion - std::max(setVersion, watchCommitVersion)